Feat/dsv 110

sammaxwellxyz opened the pull request:
Fixes #110  .

I  have since noticed there is another branch open that looks like it seeks to achieve the same goal.
`Feature110/dsv_import`

Changes proposed in this pull request:
- DSV import options
- CSV import option

Based on the CSV implementation just being a helper of the DSV it should be simple to extend for other convenience separators, e.g. pipes or semi-colons

irojkov-ph commented on src/pspm_cfg/pspm_cfg_import.m:
```diff
@@ -176,6 +176,12 @@
+delimiter.name      = 'Delimiter';
+delimiter.tag       = 'delimiter';
+delimiter.strtype   = 's';
+delimiter.help      = {['The delimiter to be used for file reading, leave blank to ']};
```
I think you forget to finish the sentence xD

sammaxwellxyz replied:
ha ha thanks

irojkov-ph commented on src/pspm_get_txt.m:
```diff
@@ -1,10 +1,11 @@
-function [sts, import, sourceinfo] = pspm_get_txt(datafile, import)
+function [sts, import, sourceinfo] = pspm_get_txt(datafile, import, delimiter)
```
Again as I pointed it out last time, I prefer to be consistent in functions names and arguments, for different data file type there should be only `datafile` and `import` as argument (look at `pspm_get_eyelink`, `pspm_get_physlog`, etc.) So I propose to put this `delimiter` option into the `import` structure, or something like that.

sammaxwellxyz replied:
I remember being conflicted on putting it in import for some reason but I'll take another look, I was likely just a bit confused

sammaxwellxyz replied:
Ah yes, so the `import` seems to me to be associated with a specific channel, not the file. Is that right? It would mean something like adding the delimiter to all import cells, then just taking the first and considering it gospel. Shall I do that? It's not particularly nice but I think that's the best way without an extra arg

sammaxwellxyz replied:
Or just put it on the first cell in the import cell array, it's hacky either way but just on first I guess is a smidgen less memory

irojkov-ph replied:
Yes that's correct, I think we can do it like for the `eyelink_trackdist` in the `pspm_get_eyelink` function where we are including it in every channel, it's a bit redundant but it's kind of logical, so what you can do is to include the delimiter in every channel of the `import` structure

irojkov-ph commented on src/pspm_import.m:
```diff
@@ -32,7 +32,9 @@
 %           - .denoise: for marker channels in CED spike format (recorded 
 %              as 'level'), filters out markers duration longer than the 
-%              value given here (in ms) 
+%              value given here (in ms)
+%           - .delimiter: for delimiter separated values, value used as delimiter for file read
+% 
```
If in `pspm_import` you put the `delimiter` in the `import` structure, why in `pspm_get_txt` you separate it from the structure ? (comment related to the comment in `pspm_get_txt`)

irojkov-ph replied:
:+1:

irojkov-ph commented on test/pspm_get_superclass.m:
```diff
@@ -47,7 +47,12 @@ function valid_datafile(this)
             
             for k = 1:numel(this.testcases)
-                [sts, import, sourceinfo] = this.fhandle(this.testcases{k}.pth, this.testcases{k}.import);
+
+                if (isfield(this.testcases{k}, 'extra_args'))
+                    [sts, import, sourceinfo] = this.fhandle(this.testcases{k}.pth, this.testcases{k}.import,this.testcases{k}.extra_args{1});
+                else;
+                    [sts, import, sourceinfo] = this.fhandle(this.testcases{k}.pth, this.testcases{k}.import);
+                end;
```
What does this do ?

sammaxwellxyz replied:
This is because I changed the function signature of the pspm_get_text so I needed the superclass to be able to handle extra args. But should be able to remove once I make the pspm_get_text the same as the others again

irojkov-ph commented on test/pspm_get_txt_test.m:
```diff
@@ -45,6 +45,48 @@ function define_testcases(this)
+            %--------------------------------------------------------------
+            this.testcases{3}.pth = 'testdatafile132435.csv';
+            
+            this.testcases{3}.import{1} = struct('type', 'scr'   , 'channel', 0, 'sr', 100);
+            this.testcases{3}.import{2} = struct('type', 'ecg'   , 'channel', 0, 'sr', 100);
+            this.testcases{3}.import{3} = struct('type', 'hr'    , 'channel', 0, 'sr', 100);
+            this.testcases{3}.import{4} = struct('type', 'resp'  , 'channel', 0, 'sr', 100);
+            this.testcases{3}.extra_args = { ',' };
+            
+            %generate testdata
+            header = {'scr' 'ecg' 'heart' 'resp'};
+            data = rand(900, 4);
+            
+            fid = fopen(this.testcases{3}.pth, 'w');
+            fprintf(fid,'scr,ecg,rate,resp\n');
+            for k=1:size(data,1)
+                fprintf(fid,'%f,%f,%f,%f\n', data(k,1), data(k,2), data(k,3), data(k,4));
+            end
+            fclose(fid);
+
+            %testcase 4 (delimiter separated value with custom delimiter (|))
+            %--------------------------------------------------------------
+            this.testcases{4}.pth = 'testdatafile132435.psv';
+            
+            this.testcases{4}.import{1} = struct('type', 'scr'   , 'channel', 0, 'sr', 100);
+            this.testcases{4}.import{2} = struct('type', 'ecg'   , 'channel', 0, 'sr', 100);
+            this.testcases{4}.import{3} = struct('type', 'hr'    , 'channel', 0, 'sr', 100);
+            this.testcases{4}.import{4} = struct('type', 'resp'  , 'channel', 0, 'sr', 100);
+            this.testcases{4}.extra_args = { '|' };
+
+            %generate testdata
+            header = {'scr' 'ecg' 'heart' 'resp'};
+            data = rand(900, 4);
+            
+            fid = fopen(this.testcases{4}.pth, 'w');
+            fprintf(fid,'scr|ecg|rate|resp\n');
+            for k=1:size(data,1)
+                fprintf(fid,'%f|%f|%f|%f\n', data(k,1), data(k,2), data(k,3), data(k,4));
+            end
+            fclose(fid);
```
Nice :+1:

irojkov-ph reviewed (changes requested):
In general good, I haven't tested it with the files you've specified but looks good :+1: !
Just few more comments:
- the testing files should be deleted from the folder (I guess you know)
- could you also add these changes in the Manual and the `release_notes.tex`, I think it could be nice to already publish it in the next release :smiley:

sammaxwellxyz commented:
@irojkov-ph thanks! all points should now be addressed.
- Removed unnecessary test files (they're generated in tests)
- added release notes
- used delimiter on import structure
- made pspm_get_csv a separate function instead of partially application

irojkov-ph commented on src/pspm_get_csv.m:
```diff
@@ -0,0 +1,19 @@
+function [sts, import, sourceinfo] = pspm_get_csv(datafile, import)
```
Great :+1:

irojkov-ph reviewed (commented):
Looks good to me :ok_hand:
